src/sentry/sentry_apps/tasks/sentry_apps.py

```diff
@@ -313,6 +313,28 @@ def _process_resource_change(
                 data[name] = serialize(instance)
 
             for installation in installations:
+                try:
+                    service_hook = ServiceHook.objects.get(
+                        organization_id=org.id,
+                        installation_id=installation.id,
+                    )
+                except ServiceHook.DoesNotExist:
+                    logger.info(
+                        "send_webhooks.missing_servicehook",
+                        extra={"installation_id": installation.id, "event": event},
+                    )
+                    continue
+
+                # Check if service hook is limited to specific projects.
+                # (applies to all projects if there are no ServiceHookProject records)
+                if ServiceHookProject.objects.filter(service_hook_id=service_hook.id).exists():
+                    project_allowed = ServiceHookProject.objects.filter(
+                        service_hook_id=service_hook.id, project_id=instance.project_id
+                    ).exists()
```

Comment:
I think it might be a good idea to cache here. Mostly these service hooks will change rarely, so we can cache the results and invalidate them when the models are updated.

Here's an example of how we do this https://github.com/getsentry/sentry/blob/c9edcbb5f1e29cc264cd8a47e8bfc30dfd6bd908/src/sentry/uptime/models.py#L270-L279

Reply:
will `@cache_func_for_models` automatically invalidate or do i need to do something else?

Reply:
It invalidates using post save hooks based on the models you give it in the decorator. So it should do it automatically if you set up the models right

Reply:
done

NOTE: had to be implemented in a very particular way to be correct

also added caching for ServiceHook that's fetched twice now